File: frontend/src/features/comunicacion/types/comunicacion.ts
/**
 * Types for the comunicacion feature.
 * Mirrors ComunicacionResponse, LoteStatusResponse, PreviewResponse from C-12 backend.
 */

export type ComunicacionEstado =
  | 'Pendiente'
  | 'Enviando'
  | 'Enviado'
  | 'Error'
  | 'Cancelado';

/** Terminal states — polling stops when all messages reach one of these. */
export const TERMINAL_ESTADOS: ComunicacionEstado[] = ['Enviado', 'Error', 'Cancelado'];

export interface ComunicacionCreate {
  asunto: string;
  cuerpo: string;
  destinatario: string;
  lote_id?: string | null;
}

export interface ComunicacionResponse {
  id: string;
  tenant_id: string;
  asunto: string;
  cuerpo: string;
  destinatario: string;
  estado: ComunicacionEstado;
  lote_id: string | null;
  error_detail: string | null;
  enviado_at: string | null;
  retry_count: number;
  created_at: string;
  updated_at: string;
}

export interface LoteStatusResponse {
  lote_id: string;
  tenant_id: string;
  total: number;
  pendientes: number;
  enviando: number;
  enviados: number;
  errores: number;
  cancelados: number;
}

export interface PreviewRequest {
  asunto: string;
  cuerpo: string;
  destinatario: string;
}

export interface PreviewResponse {
  asunto: string;
  cuerpo: string;
  destinatario: string;
}

File: backend/app/modules/comunicacion/schemas/comunicacion.py
"""Pydantic schemas for Comunicacion."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, Field

from app.modules.comunicacion.models.comunicacion import (
    ComunicacionEstado,
    InvalidStateTransitionError,
)


class ComunicacionCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    asunto: str = Field(..., min_length=1, max_length=500)
    cuerpo: str = Field(..., min_length=1)
    destinatario: EmailStr
    lote_id: UUID | None = None


class ComunicacionUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    asunto: str | None = Field(None, min_length=1, max_length=500)
    cuerpo: str | None = Field(None, min_length=1)
    destinatario: EmailStr | None = None


class ComunicacionResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: UUID
    tenant_id: UUID
    asunto: str
    cuerpo: str
    destinatario: str
    estado: ComunicacionEstado
    lote_id: UUID | None
    error_detail: str | None
    enviado_at: datetime | None
    retry_count: int
    created_at: datetime
    updated_at: datetime


class ComunicacionStateTransition(BaseModel):
    model_config = ConfigDict(extra="forbid")

    estado: ComunicacionEstado

    def validate_transition(self, current: ComunicacionEstado) -> None:
        InvalidStateTransitionError  # noqa: B018
        from app.modules.comunicacion.models.comunicacion import validate_transition
        validate_transition(current, self.estado)


class LoteStatusResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    lote_id: UUID
    tenant_id: UUID
    total: int
    pendientes: int
    enviando: int
    enviados: int
    errores: int
    cancelados: int


class LotePendienteResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    lote_id: UUID
    tenant_id: UUID
    total: int
    pendientes: int
    enviando: int
    enviados: int
    errores: int
    cancelados: int
    asunto: str
    cuerpo: str
    destinatarios: list[str]
